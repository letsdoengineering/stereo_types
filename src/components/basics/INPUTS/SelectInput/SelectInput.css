.rfs-select-container {
    width: 100%;
}

/* NOTE: important is used here due to the components default styles over-riding these otherwise */
.rfs-control-container {
    min-height: 0 !important;
    border: none !important;
    border-radius: 0 !important;
    color: var(--slate-gray) !important;
    font-size: 1rem !important;
    cursor: pointer !important;
    box-shadow: none !important;
}

/*.rfs-control-container:focus { !* Need to address the focus style at some point. */
/*    outline: 5px auto Highlight;*/
/*    outline: 5px auto -webkit-focus-ring-color;*/
/*} */

.rfs-control-container div[class^="Separator-"] {
    width: 0;
}

.rfs-control-container div[class^="ValueWrapper-"] {
    padding: 0;
}

.rfs-menu-container {
    left: calc(-1rem - 1px) !important;
    margin-top: 1.5rem !important;
    border: 1px solid var(--outline-gray) !important;
    border-radius: 0 !important;
    color: var(--slate-gray) !important;
    font-size: 1rem !important;
    box-shadow: none !important;
    cursor: pointer !important;
}

.rfs-option {
    cursor: pointer !important;
}

.rfs-option:hover {
    background-color: var(--highlight-gray)  !important;  /* Light colour for hover/focus so the selected option is clear... check with Design. */
}
.rfs-option.rfs-option-focused {
    background-color: var(--highlight-gray)  !important;  /* Light colour for hover/focus so the selected option is clear... check with Design. */
}

.rfs-option.rfs-option-selected {
    color: var(--slate-gray) !important; /* */
    background-color: var(--outline-gray) !important;
}

/** available classes on react-functional-select
rfs-select-container
rfs-control-container
rfs-menu-container
rfs-autosize-input
rfs-caret-icon
rfs-clear-icon
rfs-loading-dots
rfs-option, rfs-option-focused, rfs-option-selected, rfs-option-disabled
 */

/* DEFAULTS values, override using theme, but can not use a theme while applying ccs to classes as above: */
/* https://based-ghost.github.io/react-functional-select/?path=/story/react-functional-select--styling */
