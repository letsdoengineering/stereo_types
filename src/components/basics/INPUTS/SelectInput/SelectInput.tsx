import React, { FunctionComponent, ReactNode } from 'react'
import { Select } from 'react-functional-select'
import Icon from '../../Icon/Icon'
import './SelectInput.css'
import * as self from './SelectInput'

export type SelectOption = Readonly<{
  label: string
  value: Record<string, string> | string | null
}>

export type SelectInputProps = Readonly<{
  /** id assigned to input field */
  id?: string | undefined
  /** pass value if to start with option already selected */
  initialValue?: SelectOption | null
  /** toogle X button to clear*/
  isClearable?: boolean
  /** toggle input being interactive and acting as filter on options menu */
  isSearchable?: boolean
  /** callback that is passed the value on change events */
  onChange(data: Record<string, string>): void
  /** option that create the labels/value for the menu */
  options: SelectOption[]
  /** can provide placeholder text when no option selected - default is 'Select option..' */
  placeholder?: string
}>

/** Provides the custom icon for Select input, relies on menuOpen variable to switch between up  and down */
export const getCaretIcon = ({ menuOpen }: Record<string, any>): ReactNode => {
  const type = menuOpen ? 'chevron-up' : 'chevron-down'
  return <Icon icon={type} />
}

const SelectInput: FunctionComponent<SelectInputProps> = ({
  id,
  initialValue = null,
  isClearable = false,
  isSearchable = false,
  options,
  placeholder = undefined,
  onChange,
}) => {
  return (
    <Select
      placeholder={placeholder}
      initialValue={initialValue}
      isClearable={isClearable}
      isSearchable={isSearchable}
      options={options}
      getOptionValue={(option: SelectOption): string => option.label}
      caretIcon={self.getCaretIcon}
      onOptionChange={onChange}
      inputId={id}
    />
  )
}

SelectInput.displayName = 'SelectInput'

export default SelectInput
