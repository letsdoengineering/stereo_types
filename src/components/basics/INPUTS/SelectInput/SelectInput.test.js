import React from 'react'
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { axe } from 'jest-axe'
import * as self from './SelectInput'

const { default: SelectInput } = self
const getIconSpy = jest.spyOn(self, 'getCaretIcon')

describe('SelectInput', () => {
  const options = [
    { label: 'Any', value: 'any' },
    { label: '1-3', value: '1-3' },
    { label: '3-6', value: '3-6' },
  ]

  it('renders labeled input that shows an initial value when initialValue prop is provided', () => {
    const initialValueProp = 'my initial value'
    render(
      <>
        <label htmlFor={'an id'}>A select input</label>
        <SelectInput
          id={'an id'}
          options={options}
          initialValue={{
            label: initialValueProp,
            value: initialValueProp,
          }}
        />
      </>
    )
    const theLabel = screen.getByLabelText('A select input')
    const selectInput = screen.getByText(initialValueProp)

    expect(theLabel).toBeInTheDocument()
    expect(selectInput).toBeInTheDocument()
  })

  it('renders input that when clicked shows menu list matching the options passed in', () => {
    const onChange = jest.fn()
    render(
      <>
        <label htmlFor={'an id'}>A select input</label>
        <SelectInput id={'an id'} options={options} onChange={onChange} />
      </>
    )
    const selectInput = screen.getByLabelText('A select input')
    expect(selectInput).toBeInTheDocument()
    expect(screen.getByRole('combobox')).toHaveAttribute('aria-expanded', 'false')
    expect(getIconSpy).toHaveBeenCalledWith({ menuOpen: false })

    userEvent.click(screen.getByText('Select option..'))
    expect(screen.getByRole('combobox')).toHaveAttribute('aria-expanded', 'true')
    expect(getIconSpy).toHaveBeenCalledWith({ menuOpen: true })
  })

  it('when an option is clicked it calls onOptionChange with the value of clicked option', () => {
    const onChange = jest.fn()
    render(
      <>
        <label htmlFor={'an id'}>A select input</label>
        <SelectInput id={'an id'} options={options} onChange={onChange} />
      </>
    )
    const selectInput = screen.getByLabelText('A select input')
    expect(selectInput).toBeInTheDocument()
    expect(screen.getByRole('combobox')).toHaveAttribute('aria-expanded', 'false')
    expect(getIconSpy).toHaveBeenCalledWith({ menuOpen: false })

    userEvent.click(screen.getByText('Select option..'))
    expect(screen.getByRole('combobox')).toHaveAttribute('aria-expanded', 'true')
    expect(getIconSpy).toHaveBeenCalledWith({ menuOpen: true })

    const option1 = screen.getByText(options[0].label)
    userEvent.click(option1)
    userEvent.tab()
    expect(getIconSpy).toHaveBeenCalledWith({ menuOpen: false })
    expect(onChange).toBeCalledWith(options[0])
  })

  it('has not changed the JSX layout of the page from previous snapshot', () => {
    const tree = renderer.create(<SelectInput options={options} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders the page DOM that is accessible', async () => {
    const { container } = render(
      <>
        <label>
          a label for a select input
          <SelectInput options={options} />
        </label>
      </>
    )
    const results = await axe(container)

    expect(results).toHaveNoViolations()
  })
})
