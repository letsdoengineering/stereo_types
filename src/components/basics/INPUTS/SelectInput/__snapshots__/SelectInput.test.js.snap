// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectInput has not changed the JSX layout of the page from previous snapshot 1`] = `
<div
  aria-expanded={false}
  aria-haspopup="listbox"
  className="SelectWrapper-kcrmu9-0 dTbYzS rfs-select-container"
  onKeyDown={[Function]}
  role="combobox"
>
  <div
    className="ControlWrapper-kcrmu9-2 gJDvMS rfs-control-container"
    onMouseDown={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className="ValueWrapper-kcrmu9-1 dhLphG"
    >
      <div
        className="Placeholder-us7kwl-1 bvkXni"
      >
        Select option..
      </div>
      <input
        aria-autocomplete="list"
        autoCapitalize="none"
        autoComplete="off"
        autoCorrect="off"
        className="Input-sc-4er7q8-1 ksOaCe rfs-autosize-input"
        onBlur={[Function]}
        onFocus={[Function]}
        readOnly={true}
        required={false}
        spellCheck={false}
        style={
          Object {
            "width": 2,
          }
        }
        tabIndex={0}
        type="text"
        value=""
      />
      <div
        className="SizerDiv-sc-4er7q8-0 gXyfYO"
      >
        
      </div>
    </div>
    <div
      className="IndicatorIconsWrapper-sc-1jozl2i-0 bjIdDB"
    >
      <div
        className="Separator-sc-1jozl2i-2 jxcFlm"
        role="none"
      />
      <div
        className="IndicatorIcon-sc-1jozl2i-1 dCYoYI"
      >
        <span
          aria-hidden="true"
          className="fas fa-chevron-down iconSizeL iconColorBlue"
        />
      </div>
    </div>
  </div>
  <div
    className="MenuWrapper-sc-105ivps-0 fWmvZr rfs-menu-container"
    onMouseDown={[Function]}
  >
    <div
      onScroll={[Function]}
      style={
        Object {
          "WebkitOverflowScrolling": "touch",
          "direction": "ltr",
          "height": 0,
          "overflow": "auto",
          "position": "relative",
          "width": "100%",
          "willChange": "transform",
        }
      }
    >
      <div
        style={
          Object {
            "height": 0,
            "pointerEvents": undefined,
            "width": "100%",
          }
        }
      />
    </div>
    <div
      className="NoOptionsMsg-sc-1on2920-0 gbRWmX"
    >
      No options
    </div>
  </div>
</div>
`;
